src/compiler/generator.rs

```diff
@@ -205,4 +274,27 @@ impl Generator {
         result
     }
 
+    fn extract_expr(e: &Either<Box<NodeArithmeticOperation>, NodeBaseExpr>) -> Option<NodeArithmeticExpr> {
+        match e {
+            Right(base) => Some(NodeArithmeticExpr::Base(base.clone())),
+            Left(op) => Some(NodeArithmeticExpr::Operation((**op).clone())),
+        }
+    }
+    
+    fn type_check_logical_operands(&self, lhs: &NodeArithmeticExpr, rhs: &NodeArithmeticExpr) -> Result<(), String> {
+        if self.infer_type(lhs) == "bool" && self.infer_type(rhs) == "bool" {
+            Ok(())
+        } else {
+            Err("Logical operators can only be applied to booleans".to_string())
+        }
+    }
+    
+    fn infer_type(&self, expr: &NodeArithmeticExpr) -> &str {
+        match expr {
+            NodeArithmeticExpr::Base(NodeBaseExpr::Bool(_)) => "bool",
+            NodeArithmeticExpr::Base(NodeBaseExpr::Num(_)) => "num",
+            NodeArithmeticExpr::Base(NodeBaseExpr::ID(_)) => "unknown",
```

Comment:
We might want to start and modify the id map, so it contains the proper variable type